Add initial Tracing with OpenTelemetry

lilic opened the pull request:
This PR adds initial tracing using [OpenTelemetry](https://github.com/open-telemetry), as per discussion on the https://github.com/etcd-io/etcd/issues/12460. Let me know if we want this merged in 3.5, I believe it still adds value and observability with some tradeoffs to resource usage, that I detailed below.

## This introduces:
- Setups up OpenTelemetry tracing.
- New experimental flags for Tracing which for enabling tracing, specifying OpenTelemetry collector address and customizing service name. Usage:
`etcd --experimental-enable-tracing --experimental-tracing-address="0.0.0.0:55680" --experimental-tracing-service-name="etcd-new"` 
- Only if the `experimental-enable-tracing` is enabled does tracing of the gRPC start collecting.
- This is how the traces look like visulazed in jaeger UI:
<img width="1431" alt="Screenshot 2021-05-05 at 12 19 55" src="https://user-images.githubusercontent.com/6232346/117163974-76027580-adc4-11eb-8db7-dd17848dec3e.png">
And the individual gRPC request span:
<img width="1434" alt="Screenshot 2021-05-05 at 11 09 54" src="https://user-images.githubusercontent.com/6232346/117164419-e4473800-adc4-11eb-8e3f-46a25ef429cf.png">

## Performance/resource usage

I ran an initial small load with tracing on and found that overall on average it adds a 1.5% - 4% CPU overhead:

This is the merged profile during the total load test: https://share.polarsignals.com/d92bfce/, we can see under the icicle graph there that the `otelgrpc.UnaryServerInterceptor` function was 1.84% of the total CPU. Plus extra 0.84% CPU time in the` otelgrpc.messageType.Event`.

In some individual profiles, we can see around 4.71% CPU overhead. https://share.polarsignals.com/e5d524c/ 

Happy to share more profiles, if needed, or run more load tests!

## Open Questions

1. Do we need to add tracing to the grpc-proxy as well?
2. OpenTelemetry go client is not yet GA or stable, between the last two minor releases there were some breaking changes, is that okay for etcd? Can we bump the version of this in a patch etcd release or after code freeze?

## Follow up improvements and features:
- I will add some documentation if we decide to merge this in a follow-up PR. And mention the above resource overhead to users.
- We can add more finite traces for various parts of the codebase in the future.
- More configuration options in the future, like sampling configuration.

lilic commented:
@ptabor @dashpole @logicalhan @hexfusion please have a look, thank you!

lilic commented on go.sum:
```diff
@@ -294,9 +301,33 @@ go.etcd.io/bbolt v1.3.5 h1:XAzx9gjCb0Rxj7EoqcClPD1d5ZBxZJk0jbuoPHenBt0=
 go.etcd.io/bbolt v1.3.5/go.mod h1:G5EMThwa9y8QZGBClrRx5EY+Yw9kAhnjy3bSjsnlVTQ=
 go.opencensus.io v0.21.0/go.mod h1:mSImk1erAIZhrmZN+AvHh14ztQfjbGwt4TtuofqLduU=
 go.opencensus.io v0.22.0/go.mod h1:+kGneAE2xo2IficOXnaByMWTGM9T73dGwxeWcUqIpI8=
+go.opentelemetry.io/contrib v0.20.0 h1:ubFQUn0VCZ0gPwIoJfBJVpeBlyRMxu8Mm/huKWYd9p0=
```
I am not sure why all the imports are there, I tried running `go mod tidy` but did not help, but I guess internally opentelemetry relies on all of these?

dashpole replied:
So... OpenTelemetry wants to make the modules related to tracing go to 1.0 without making the modules related to metrics go 1.0.  They decided to make smaller modules so to make that possible.  So I think the moldules below are correct.

dims commented:
@lilic just to be sure zero cost if experimental-enable-tracing is false (which is default) right?

hexfusion reviewed (commented):
Thanks, @lilic overall I think this is a nice feature to try and move forward into 3.5. apiserver[1] is working on this as well so the pairing will be an important part of the observability story for k8s. While the performance hit is unfortunate documenting it as such and gating under an experimental flag seems appropriate. My understanding is that there are continuing efforts to improve performance. +1

[1] https://github.com/kubernetes/kubernetes/pull/94942

bboreham commented on server/go.mod:
```diff
@@ -33,14 +33,19 @@ require (
 	golang.org/x/time v0.0.0-20210220033141-f8bda1e9f3ba
 	google.golang.org/genproto v0.0.0-20200526211855-cb27e3aa2013
-	google.golang.org/grpc v1.36.1
+	google.golang.org/grpc v1.37.0
```
In another project I switched from otelgrpc to otelhttp to avoid this problem.

lilic replied:
Thanks Bryan I will look into that, if we don't want to make this jump!

hexfusion replied:
as per conversation at the community meeting, there are no known breaking changes moving to 1.37

dims commented on server/embed/etcd.go:
```diff
@@ -770,3 +794,24 @@ func parseCompactionRetention(mode, retention string) (ret time.Duration, err er
+	if err != nil {
+		return nil, err
+	}
+	res := resource.NewWithAttributes(
```
is `ExperimentalTracingServiceName` mandatory? can it be empty?

lilic replied:
It can be empty, it has a default value of `"etcd"`. But I would say we should recommend to users to make use of it, for the reason detailed in https://github.com/etcd-io/etcd/pull/12919#discussion_r627103132.

The only thing you need to pass to enable tracing with OpenTelemetry is the enable flag `etcd --experimental-enable-tracing`.

dashpole replied:
Setting service name differently for each instance of a horizontally scaled service goes against the Otel conventions: https://github.com/open-telemetry/opentelemetry-specification/blob/e440dd50a16f4ff4f5ff47e420a49129127f7a37/specification/resource/semantic_conventions/README.md#service.  We should use other attributes (e.g. service.instance.id) to differentiate instances of etcd from each other.  Service.name should be a constant.

lilic replied:
Interesting approach, does this get translated into the tag in jaeger UI?

dashpole replied:
Yes.  Attributes are just key-value pairs, so you can add whatever attributes you want to, and they will show up in the Jaeger UI.  In general, we should try to use semantic conventions whenever possible, but if we have our own attribute that we don't think maps well to existing conventions (e.g. etcd replica number), we can just add our own here.

lilic replied:
Added ability of setting those now, but haven't found a way to filter by them yet in the jaeger UI. As it shows up under special filed not under tags for me.

But good suggestion!

lilic commented on server/embed/config.go:
```diff
@@ -308,6 +308,16 @@ type Config struct {
+	ExperimentalTracingAddress string `json:"experimental-tracing-address"`
+	// ExperimentalTracingServiceName is the name of the service.
+	// Can only be used if ExperimentalEnabledTracing is true.
+	ExperimentalTracingServiceName string `json:"experimental-tracing-service-name"`
```
Having the ability to specify the name allows you to collect the traces from all the running instances of etcd separately, so you can detect issues per running instance of etcd, this way you can in your UI select just the instance you had an issue with and troubleshoot.

gyuho replied:
> Having the ability to specify the name allows you to collect the traces from all the running instances of etcd separately, so you can detect issues per running instance of etcd, this way you can in your UI select just the instance you had an issue with and troubleshoot.

Can we have this as flag description or some format of doc :) 

e.g.,

```
// ExperimentalTracingServiceName is the tracing service name for
// OpenTelemetry (if enabled) -- "etcd" is the default service name.
// When shared, all telemetry data are aggregated under the same namespace.
// Use different names in order to collect data per each node.
```

(if I understand your comment correctly)

lilic replied:
That is a good idea, will add it there!

The plan was to also add some docs if PR is merged, maybe similar to where metrics docs are right now.

hexfusion replied:
> Having the ability to specify the name allows you to collect the traces from all the running instances of etcd separately, so you can detect issues per running instance of etcd, this way you can in your UI select just the instance you had an issue with and troubleshoot.

If I understand what your saying this name should be unique per etcd? This is why I was wondering if using etcd member name[1] which is unique for the cluster would make sense maybe as a default.

[1]https://github.com/etcd-io/etcd/blob/master/server/config/config.go#L37

logicalhan replied:
Oh I like that idea.

lilic replied:
Yes should be unique per etcd process/member.

Missed this comment sorry, nice will use that as default! But I think we should still let users override it.

lilic replied:
On second thought, if the name is not passed it defaults to... `default`. IMHO this makes the discoverability out of the box in the tracing UI a bit harder, when the service is called `default`. Should we stick to `Name` as default or constant`"etcd"`?

hexfusion replied:
how about if the name is `default` use `etcd`  otherwise use name and allow override with flag?

lilic commented:
> just to be sure zero cost if experimental-enable-tracing is false (which is default) right?

Exactly, we only ever enable OTel tracing if that flag is passed yes.

You can see this more clearly here https://share.polarsignals.com/442c0d2/ I ran some load when the `experimental-enable-tracing` flag was passed and the same load when the flag was not enabled, you can see the CPU diff if you look at the top view and search for telemetry.

gyuho commented on server/config/config.go:
```diff
@@ -163,6 +163,9 @@ type ServerConfig struct {
 	EnableGRPCGateway bool
 
+	// ExperimentalEnabledTracing enables tracing via OpenTelemetry.
+	ExperimentalEnabledTracing bool
```
`s/Enabled/Enable/`?

logicalhan replied:
nit: I prefer EnableExperimentalTracing to be consistent with the flag above it.

lilic replied:
I was following the examples of `ExperimentalMemoryMlock` and `ExperimentalTxnModeWriteWithSharedBuffer` that exist already in the `ServerConfig `.

logicalhan replied:
ah

logicalhan replied:
maybe ExperimentalTracing then?

lilic replied:
@gyuho suggested the distributed part as etcd already has its own tracing, see this https://github.com/etcd-io/etcd/pull/12919#issuecomment-833448429

gyuho commented on server/embed/config.go:
```diff
@@ -308,6 +308,15 @@ type Config struct {
 	ListenMetricsUrlsJSON string `json:"listen-metrics-urls"`
 
+	//ExperimentalEnabledTracing indicates if experimental tracing using OpenTelemetry is enabled.
+	ExperimentalEnabledTracing bool `json:"experimental-enable-tracing"`
```
`s/Enabled/Enable/`?

gyuho commented on server/etcdmain/config.go:
```diff
@@ -249,6 +249,11 @@ func newConfig() *config {
 	fs.StringVar(&cfg.ec.Metrics, "metrics", cfg.ec.Metrics, "Set level of detail for exported metrics, specify 'extensive' to include server side grpc histogram metrics")
 
+	// experimental tracing
+	fs.BoolVar(&cfg.ec.ExperimentalEnabledTracing, "experimental-enable-tracing", false, "Enable experimental tracing using OpenTelemetry Tracing.")
```
Can we define defaults as public const somewhere?

gyuho reviewed (changes requested):
Looks great overall. Requesting some minor changes. Thanks!

gyuho commented:
@lilic On second thought. tracing is not descriptive enough, given that we do already have tracing within etcd process. I think we should name it something like `distributed-tracing`?

lilic commented:
I think that is a good idea yes, for clarity! So we would have everything prefixed with `experimental-distributed-tracing-*`.

logicalhan commented on server/embed/etcd.go:
```diff
@@ -249,6 +260,14 @@ func StartEtcd(inCfg *Config) (e *Etcd, err error) {
 	}
 
+	if e.cfg.ExperimentalEnabledTracing {
+		e.tctx = context.Background()
```
I believe it is recommended not to do this: (fifth paragraph: https://golang.org/pkg/context/ )

It's probably better form to pass context directly into setupTracing. 

Also, why not invoke the defer tracingExporter.Shutdown right here (but before the conditional error return) on like 268?

lilic replied:
Because defer gets called straight away as the `StartEtcd` func returns. So I placed it in the `Close` where the `metricsListeners` is also shut down.

> // Close gracefully shuts down all servers/listeners.

But happy to adapt this differently if you have any other suggestions!

Will pass in context directly, wasn't too happy with that either you are right!

logicalhan replied:
Oh i didn't realize it was in a different function, the diff obscured that from me. I guess you'd have to wrap a lambda function for shutdown so you can pass in the context here, that should allow you to get away from storing context.

something like

```
   ctx := context.Background
   \\ tracing initialization
   e.shutdownTracing := function() {
     e.e.tracingExporter.Shutdown(ctx)
   }
```

lilic replied:
Makes sense, will do that. Was just following the current pattern of saving context in the struct that exists, but agreed on this approach!

lilic replied:
Done, please have a look thanks!

lilic commented on server/embed/config.go:
```diff
@@ -68,6 +68,11 @@ const (
 	StdErrLogOutput  = "stderr"
 	StdOutLogOutput  = "stdout"
 
+	// ExperimentalDistributedTracingAddress is the default collector address.
```
I wasn't sure if this is the right place to put the constants? @gyuho any suggestions?

gyuho replied:
Yes, here's good. Thanks!

lilic commented:
I noticed the help text file (server/etcdmain/help.go) , is there a way to generate that file or is it manually edited?

logicalhan reviewed (commented):
lgtm :)

gyuho commented:
> I noticed the help text file (server/etcdmain/help.go) , is there a way to generate that file or is it manually edited?

We manually edit :)

gyuho reviewed (approved):
lgtm, thanks!

dashpole commented on server/embed/etcd.go:
```diff
@@ -774,3 +797,29 @@ func parseCompactionRetention(mode, retention string) (ret time.Duration, err er
+		tracesdk.WithSyncer(exporter),
+		tracesdk.WithResource(res),
+	)
+	otel.SetTracerProvider(tp)
```
Just to call this out to reviewers less familiar with opentelemetry: This is setting a global variable, which is used by tracing libraries.  In the kubernetes APIServer, we will likely avoid doing this by explicitly passing the TracerProvider to the libraries we want to do tracing.  It is more code, but avoids global variables.

lilic replied:
Thanks! Yeah right now we are not adding any spans only using the gRPC so I don't see it being different, but I could be wrong?

dashpole replied:
If, for example, a library you rely on added OpenTelemetry tracing, you would start getting traces from that library without changing anything on your end, since it will look up the global TracerProvider to see where to send traces.  That may or may not be desirable.  Also, if you have multiple different "services" in the same go binary, you can't customize the TracerProviders individually for each service while using the global TracerProvider.  I think it is perfectly fine to start with using the global one, though.

ptabor replied:
Embed server is being used 'as a library' by our customers. Also in tests we create multiple instances of Embed server to mimic multiple etcd nodes. Recently we get rid of 'global loggers'  (including grpc loggers) being overridden by each of the embed servers (leading to misleading zap-fields being reported). 

I don't know how differently `otl` can be configured for different libs, but  having lib-specific tracer seems to be on the safer side.

lilic replied:
I am all for removing globals, I felt the pain many times with Prometheus global registry (which I would love to remove from etcd one day as well, if folks agree). 

From what I can tell we can pass it via https://pkg.go.dev/go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc#WithTracerProvider, so will do that.

dashpole replied:
Agreed.  I opened https://github.com/open-telemetry/opentelemetry-go/issues/1888

Aneurysm9 replied:
> Embed server is being used 'as a library' by our customers. Also in tests we create multiple instances of Embed server to mimic multiple etcd nodes.

If the embedded server is used as a library it may be better to avoid use of the OTel SDK at all and only use the API.  In that case the server should receive its TracerProvider and propagators using the API interfaces from the `main` package that has instantiated them with the SDK.  That would allow the consumer of the embedded server to decide whether to enable tracing, which exporters and propagators to use, even which SDK to use if alternate implementations of the SDK become available.

dashpole commented on server/embed/etcd.go:
```diff
@@ -774,3 +797,29 @@ func parseCompactionRetention(mode, retention string) (ret time.Duration, err er
+		semconv.ServiceNameKey.String(e.cfg.ExperimentalDistributedTracingServiceName),
+	)
+	tp := tracesdk.NewTracerProvider(
+		tracesdk.WithSyncer(exporter),
```
WithSyncer is synchronous (blocks until the trace has been sent).  See https://github.com/open-telemetry/opentelemetry-go/blob/24a91628f112f234af024afb42be57247752bbd0/sdk/trace/provider.go#L240.  You should use WithBatcher instead.

dashpole commented on server/embed/etcd.go:
```diff
@@ -774,3 +797,29 @@ func parseCompactionRetention(mode, retention string) (ret time.Duration, err er
+		tracesdk.WithBatcher(exporter),
+		tracesdk.WithResource(res),
+	)
+	otel.SetTracerProvider(tp)
```
in addition to setting the global TracerProvider, you should also set the global propagators:

```golang
otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(propagation.TraceContext{}, propagation.Baggage{}))
```

The TraceContext propagator will allow linking traces from the APIServer to etcd, and Baggage allows passing additional tags via context.

dashpole commented:
One thing you don't do here, but probably should is add a sampling policy.  The default is to ParentBased(AlwaysSample) (always sample unless there is an incoming parent context that is explicitly not sampled), which is quite verbose.  Something like sampling 1% of requests is probably more appropriate.  But that can easily be done in a follow-up

lilic commented:
> But that can easily be done in a follow-up

Yes agreed, this was a minimal PR and was also one of my questions in the PR description, if we can do follow-ups even in the patch releases that would include new configurations.

ptabor commented:
The tests fail due to:
```
FAIL: inconsistent versions for depencency: github.com/golang/protobuf
  - github.com/golang/protobuf@v1.5.1 from: go.etcd.io/etcd/api/v3
  - github.com/golang/protobuf@v1.5.1 from: go.etcd.io/etcd/raft/v3
  - github.com/golang/protobuf@v1.5.2 from: go.etcd.io/etcd/server/v3
  - github.com/golang/protobuf@v1.5.2 from: go.etcd.io/etcd/tests/v3
FAIL: inconsistent versions for depencency: google.golang.org/grpc
  - go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc@v0.20.0 from: go.etcd.io/etcd/server/v3
  - google.golang.org/grpc@v1.36.1 from: go.etcd.io/etcd/api/v3
  - google.golang.org/grpc@v1.36.1 from: go.etcd.io/etcd/client/pkg/v3
  - google.golang.org/grpc@v1.36.1 from: go.etcd.io/etcd/client/v3
  - google.golang.org/grpc@v1.36.1 from: go.etcd.io/etcd/pkg/v3
  - google.golang.org/grpc@v1.37.0 from: go.etcd.io/etcd/etcdctl/v3
  - google.golang.org/grpc@v1.37.0 from: go.etcd.io/etcd/server/v3
  - google.golang.org/grpc@v1.37.0 from: go.etcd.io/etcd/tests/v3
  - google.golang.org/grpc@v1.37.0 from: go.etcd.io/etcd/v3
FAIL: inconsistent dependencies
```

Should get fixed by: 
```
./scripts/update_dep.sh  github.com/golang/protobuf v1.5.2
./scripts/update_dep.sh google.golang.org/grpc v1.37.0
```

ptabor commented on server/embed/etcd.go:
```diff
@@ -249,6 +258,15 @@ func StartEtcd(inCfg *Config) (e *Etcd, err error) {
+	if e.cfg.ExperimentalEnableDistributedTracing {
+		tctx := context.Background()
+		tracingExporter, err := e.setupTracing(tctx)
+		e.tracingExporterShutdown = func() { tracingExporter.Shutdown(tctx) }
```
Shell we register it even if e.setupTracing() returned  (nil, err) ?

lilic replied:
Good catch, I added a check for either to be nil and I return a tracing setup error. Please have another look, thanks!

lilic commented on server/etcdmain/config.go:
```diff
@@ -249,6 +249,12 @@ func newConfig() *config {
+	fs.BoolVar(&cfg.ec.ExperimentalEnableDistributedTracing, "experimental-enable-distributed-tracing", false, "Enable experimental distributed  tracing using OpenTelemetry Tracing.")
+	fs.StringVar(&cfg.ec.ExperimentalDistributedTracingAddress, "experimental-distributed-tracing-address", embed.ExperimentalDistributedTracingAddress, "Address for distributed tracing used for OpenTelemetry Tracing (if enabled with experimental-enable-distributed-tracing flag).")
+	fs.StringVar(&cfg.ec.ExperimentalDistributedTracingServiceName, "experimental-distributed-tracing-service-name", embed.ExperimentalDistributedTracingServiceName, "Configures service name for distributed tracing to be used to define service name for OpenTelemetry Tracing (if enabled with experimental-enable-distributed-tracing flag). 'etcd' is the default service name. Use the same service name f ...
+	fs.StringVar(&cfg.ec.ExperimentalDistributedTracingServiceInstanceID, "experimental-distributed-tracing-instance-id", "", "Confgiures service instance ID for distributed tracing to be used to define service instance ID key for OpenTelemetry Tracing (if enabled with experimental-enable-distributed-tracing flag). There is no default value set. This ID must be unique per etcd instance.")
```
I default to empty string here, which we check for when doing the tracing setup. If anyone has a better suggestion for default value I can change it. 

But since it's not a required field I did it this way.

hexfusion replied:
minor: `s/Confgiures/Configures`

hexfusion replied:
```
ExperimentalDistributedTracingServiceInstanceID
```
I think we should try eventually to populate this with reasonable default. etcd name or the peerURL are thoughts.

lilic replied:
Continued discussion on this below, so will resolve this one.

lilic commented:
Okay, this should be ready for another look:
- I added ability of setting service instance ID which should be the unique key, with service name being always the same (added comments)
- Removed the global setting and I instead pass the tracer around to the gRPC.
- Various other safe guard checks.

Please have another look, thank you!

TODO from my side:
- [ ] Change help text once we decide on the flag names and default values as its manual process.
- [ ] Will add some docs in a follow-up PR.

hexfusion commented on server/config/config.go:
```diff
@@ -163,6 +164,11 @@ type ServerConfig struct {
 
+	// ExperimentalEnableDistributedTracing enables distributed tracing using OpenTelemetry protocol.
+	ExperimentalEnableDistributedTracing bool
+	// ExperimentalTracerProvider is the tracer to be used when creating an OpenTelemtry Tracer.
```
minor `s/OpenTelemtry/OpenTelemetry`

hexfusion commented on server/embed/etcd.go:
```diff
@@ -774,3 +804,49 @@ func parseCompactionRetention(mode, retention string) (ret time.Duration, err er
+	)
+	// As Tracing service Instance ID must be unique, it should
+	// never use the empty default string value, so we only set it
+	// if it's a non empty string.
```
so if Instance ID is empty string is that an error?

lilic replied:
As this field is optional. We just don't add it, as that is our default value, but also as an empty string can never be unique and the ID must be unique. Happy to do another approach here well!

hexfusion replied:
yeah I think driving reasonable defaults from the current member would be a nice option. If the flag is not required folks will leave it blank and then get errors later on. as noted above etcd name and or peerURL could be considered.

hexfusion replied:
this could be a followup

lilic replied:
etcd name is not unique per etcd instance, at least not by default (`default` :) ). But happy to do `peerURL`. 

There tend to be two approaches from what I have seen, either nothing by default or in some places, hostname is what some projects use the default.

lilic commented on server/embed/etcd.go:
```diff
@@ -774,3 +804,49 @@ func parseCompactionRetention(mode, retention string) (ret time.Duration, err er
+		tracesdk.WithResource(res),
+	)
+
+	otel.SetTextMapPropagator(
```
I am guessing we need this for apiserver (or other components that would also add this kind of tracing) as you mentioned above, or is there a way to avoid that?

dashpole replied:
You can plumb it through to the interceptors, and pass it as an argument: https://github.com/open-telemetry/opentelemetry-go-contrib/blob/13d72c9c8dda39280ead2e291cbed8f51a6f65f9/instrumentation/google.golang.org/grpc/otelgrpc/grpctrace.go#L67

dashpole replied:
Basically do the same thing you did with the TracerProvider

dashpole reviewed (approved):
lgtm!

ptabor commented:
`./scripts/fix.sh` needed to refresh bom
